### PR TITLE
Chitter: kind of, testing: kind of, async functions: what? Jack M.

### PR DESCRIPTION
**Pull request by jackmcc08**
Please see Readme for updates on what I did. 

This was a tough one :D

### CODE REVIEW

File: DuckTestRunner.html [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  <script src="lib/isItDuckingWorking/testFunction.js" defer></script>
+  <script src="spec/interfaceDuckSpec.js" defer></script>
+
+  <title>Is it Ducking Working?</title>
+
+<!-- Image credit: giphy.com - EASPORTSNHL -->
+  </head>
+  <body>
+    <div style="width:100%;height:0;padding-bottom:100%;position:relative;"><iframe src="https://giphy.com/embed/cmjxqGDaVTH3apOeLa" width="100%" height="100%" style="position:absolute" frameBorder="0" class="giphy-embed" allowFullScreen></iframe></div><p><a href="https://giphy.com/gifs/cmjxqGDaVTH3apOeLa">via GIPHY</a></p>
+  </body>
```
Comment (StuBehan) [inline DuckTestRunner.html:20]: Noice.

File: src/interface.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+          <p id='time_text'>${cryTime}</p>
+        </div>
+        <div id='cry_content_div'>
+          <p id='cry_content'>${cry.body}</p>
+        </div>
+      </div>
+      `
+      cryerContainer.innerHTML += cryDisplay
+    });
+  }
```
Comment (StuBehan) [inline src/interface.js:67]: this works really well on the index page.

File: spec/interfaceJasmineSpec.js [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+      expect(testCryDiv.children[0].children[0].children[0].innerText).toEqual(`kay`);
+      expect(testCryDiv.children[0].children[0].children[1].innerText).toEqual(`19:48, 16-02-2021`);
+      expect(testCryDiv.children[0].children[1].innerText).toEqual("This is a cry!");
+      expect(testCryDiv.children[1].children[1].innerText).toEqual("This is another cry!");
+
+      //reset
+      document.getElementById("cryer-div").remove();
+      document.getElementById("welcome-wagon-div").remove();
+    })
+  })
```
Comment (StuBehan) [inline spec/interfaceJasmineSpec.js:67]: This is a solid test, I'm able to understand what is happening without much effort.

File: spec/interfaceJasmineSpec.js [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        <input type="text" placeholder="password" id="password-entry" required>
+        <input type="submit" value="Submit">
+      </form>
+      <form id="cry-entry-box" hidden="true">
+        <textarea id='cryArea' rows="15" cols="42" placeholder="Write your Cry here!"></textarea>
+        <input type="submit" value="Submit">
+      </form>
+      </div>`
+    )
+  })
```
Comment (StuBehan) [inline spec/interfaceJasmineSpec.js:35]: Good setup for your tests, was this written before the page was set up? à la TDD?

**Review by StuBehan — COMMENTED** [review]
I think this is a solid effort and the functionality that is there is really good.

File: src/interface.js [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        document.getElementById("log-in").hidden = false
+      } else if (output.handle[0] === "has already been taken") {
+        console.log("This username is already taken.");
+        window.alert(`Username already taken, please try again.`)
+        document.getElementById("password-entry").value = ''
+      } else {
+        console.log("unknown error");
+        window.alert(`An unknown error has occured, please contact your adminstrator.`)
+      }
+      console.log(output)
```
Comment (siellsiell) [inline src/interface.js:88]: You asked
> To be more specific, in my app, when signing up a new user submitNewUser, I wanted the return from the API to be returned out of the Interface class into the app.js  signupForm event listener so that I could then manipulate the HTML on that basis. But I could not get the variables in the app.js to wait for the data to be fetched from the API. How should the receivers be set up?

The code inside the callback function I highlighted could instead be defined inside a function that lives in `app.js`. This function can then passed in to `submitNewUser` as a callback when you call it `submitNewUser` from `app.js`. So it would look kind of like this:


```
// app.js

signupButton.addEventListener('click', function() {
  // other stuff ...
  theInterface.submitNewUser(inputString, handleNewUser)
}

function handleNewUser(newUserResponse) {
    // manipulate the DOM based on the response
}  
```

```
// interface.js
submitNewUser(handlePassword, newUserCallback) {
    let loginInfo = { session: handlePassword }
    this.makePostRequest("https://chitter-backend-api-v2.herokuapp.com/sessions", loginInfo, function(output) {
      newUserCallback(output)
    }
}
```

Another, perhaps simpler option could be to make your function in the `Interface` class return a promise (you can achieve that with the `async` keyword) and then handle that promise from `app.js`:

```
// app.js
signupButton.addEventListener('click', function() {
  // other stuff ...
  theInterface.submitNewUser(inputString).then( json => {
    // update the DOM
 })
}
```

```
// interface.js

class Interface {
  async submitNewUser(handlePassword) {
       let response = await fetch("https://...");
       return response.json();
  }
}
```
I can't guarantee that all these code snippets work out of the box but that should give you an idea. This resource might help with understanding better how async and await can be used: https://javascript.info/async-await

File: src/interface.js [hunk 6]
```diff
@@ -0,0 +1,127 @@
+let sessionKey
```
Comment (siellsiell) [inline src/interface.js:1]: I'm not sure that `interface` is the right name for this file since it sounds like user interface when actually it's about interacting with the api. You could call it `api.js` instead or `api_service.js`.

You might even decompose this file further to keep things neat and more SRP-ish:
- `cryService.js` for everything to do with CRUD API operations on posts
- `usersService.js` for API calls that allow logging in, signing up, logging out

File: app.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
+let theInterface = new Interface(makeGetRequest, makePostRequest)
```
Comment (siellsiell) [inline app.js:1]: I like that you have a separate `app.js` file for dealing with the DOM and triggering API calls.

**Review by siellsiell — COMMENTED** [review]
I left a few comments in response to your request on #coaches. I haven't had time yet to look into your questions about testing/mocking. Will get back to you on that (and other open questions) tomorrow.

File: spec/interfaceDuckSpec.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  let testDiv = document.createElement("div");
+  testDiv.setAttribute('id', 'cryer-div');
+  document.body.appendChild(testDiv)
+
+  let stubbedData = [
+    { body: "This is a cry!", updated_at: "2021-02-16T19:48:17.065Z", user: {id: 34, handle: "kay"}},
+    { body: "This is another cry!", updated_at: "2021-02-16T19:48:17.065Z", user: {id: 34, handle: "kay"}}
+  ]
+  let testInterface = new Interface((url, callback) => {
+      callback(stubbedData)
```
Comment (siellsiell) [inline spec/interfaceDuckSpec.js:16]: You found an approach to inject mock data into your interface that works but the reason you had to do that in the first place is that your `Interface` also takes care of displaying things instead of only doing API-related stuff. If you manage to move all of the display related code into `app.js` with some of the approaches I suggested in another comment, you won't have to inject mock data here.

You would still have to mock your calls to `fetch` which you could do manually without the use of a mocking library using an approach like the ones in this pill: https://github.com/makersacademy/course/blob/master/pills/mocking_in_javascript.md
Basically in your tests, you would overwrite `window.fetch` with a different function in your tests.

Other JS testing libraries like Jest have some helpers to do that a bit more easily but that's the general idea.
